### PR TITLE
feat: adding implementation for transferFromWithPermit in PPO

### PR DESCRIPTION
**Pull request by Nilay27**
* Adding implementation of `transferFromWithPermit()` in `PPO.sol` along with natspecs in `IPPO.sol`.
* Also added utility function `getPermitSignature()` to generate `v,r,s` to `transferFromWithPermit()`.
* Splitting this PR into two parts : 1 for implementation and another for tests to avoid making this PR too long.

### CODE REVIEW

**Review by ramenforbreakfast — APPROVED** [review]
LGTM, please next time have at least the contract changes itself and test in the same PR, the utils and interface can be separated.

File: apps/smart-contracts/token/utils/index.ts [hunk 1]
```diff
@@ -153,3 +155,61 @@ export function generateDomainSeparator(
     SignTypedDataVersion.V4
   ).toString('hex')}`
 }
+
+// from UniswapV3 Periphery's test suite
+export async function getPermitSignature(
+  signer: SignerWithAddress,
+  token: ERC20PermitUpgradeable,
+  spender: string,
```
Comment (ramenforbreakfast) [inline apps/smart-contracts/token/utils/index.ts:163]:
```suggestion
  token: ERC20PermitUpgradeable,
  signer: SignerWithAddress,
  spender: string,
```
Confusing to have the token in between signer and spender

File: apps/smart-contracts/token/contracts/ppo/PPO.sol [hunk 2]
```diff
@@ -34,6 +34,23 @@ contract PPO is IPPO, SafeOwnableUpgradeable, ERC20BurnableUpgradeable, ERC20Per
...
   }
 
+  function transferFromWithPermit(
+    address _account,
+    address _recipient,
+    uint256 _amount,
+    uint256 _deadline,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
```
Comment (XavierEkkel) [inline apps/smart-contracts/token/contracts/ppo/PPO.sol:44]:
```suggestion
    uint8 _v,
    bytes32 _r,
    bytes32 _s
```
for consistency

**Review by XavierEkkel — APPROVED** [review]
LGTM after comments

File: apps/smart-contracts/token/contracts/ppo/PPO.sol [hunk 3]
```diff
@@ -34,6 +34,23 @@ contract PPO is IPPO, SafeOwnableUpgradeable, ERC20BurnableUpgradeable, ERC20Per
...
+  ) external override {
+    /**
+     * `permit()` gives `_msgSender()` the allowance to spend `_amount` of
+     * PPO on behalf of `_account` before `_deadline` ends.
+     */
```
Comment (XavierEkkel) [inline apps/smart-contracts/token/contracts/ppo/PPO.sol:49]:
```suggestion
```
interface documentation is enough

File: apps/smart-contracts/token/contracts/ppo/interfaces/IPPO.sol [hunk 4]
```diff
@@ -40,6 +40,26 @@ interface IPPO is IERC20Upgradeable, IERC20PermitUpgradeable {
    */
   function burnFrom(address account, uint256 amount) external;
 
+  /**
+    * @notice Transfers `amount` of token from `account` to `recipient` without 
+    * needing additional transaction for `approval()`.
+    * @dev The `deadline` must be a timestamp in the future.
```
Comment (XavierEkkel) [inline apps/smart-contracts/token/contracts/ppo/interfaces/IPPO.sol:46]: add:
`@dev `from`, `to` and `deadline` must exactly match the values used to generate `v`, `r`, `s`.

**Pull request merged**